source/_posts/2018-01-12-pager-duty-and-confluence.markdown

```diff
@@ -0,0 +1,253 @@
+---
+layout: post
+title: "Pager Duty and Confluence"
+date: 2018-01-12 18:19:49 +0000
+comments: true
+author: "Bryan Kroger bryan.kroger@grandrounds.com"
+categories: automation
+---
+My first project here at Grand Rounds was to pick up one of my favorite topics of IT and operations 
+in general: on-call rotations!  Making sure we have accurate alarming and alerting never seems like the
+most glamours work in the world, but it is often very important and highly impactful.
+
+To that end, as with most things, we want to start with a data-driven approach to identifying and resolving
+our problems.  This is a pattern that I've used several times over the years for a fun little hack which
+allows us to build reports based on PagerDuty incident data.
+
+This gives us two graphs:
+
+* Number of incidents per service.
+* Amount of time each service spends in incident ( on average ).
+
+We fully expect that this code is a little rough around the edges, and that's okay.  We're building things
+here that are going to give us ideas and insight into our bigger picture items.  Ideally we want to
+spend as little time as possible gold plating this code because we're not 100% sure this will be something
+that we intend to live forever.  Just like in the startup business world we want to create the Minimal
+Viable Product ( MVP ) which allows us to prove out a business case.
+
+In this case we will prove out the value of incident reporting data.  This is a perfect example of a product
+that is valuable today, but becomes less valuable over time because it allows us to refine a process.
+
+We fully expect that some day this data will be effictively useless and thus this code will be thrown
+away.  However, the value of this code today is in it's ability to give us insights that we can act on
+today.
+
+We start this with a high-level HTTPClient.  Our nickname for this is called Groot: 
+Grand Rounds Operational and Organizational Tool.  We're all fans of the comic books over here, 
+so we often name our projects after Marvel characters.  I haven't seen anything from the DC 
+realm quite yet, but maybe someday.
+
+```ruby
+class GrootHTTPClient
+  @url
+  @headers
+
+  def initialize
+    @url = get_url
+    @headers = get_headers
+  end
+
+  def get_headers
+    {
+        'Content-Type' => 'application/json'
+    }
+  end
+end
+```
+
+This is a very simply little base class to get us started.  Now let's build a specific client on top of this:
+
+```ruby
+class ConfluenceClient < GrootHTTPClient
+  @username
+  @password
+
+  def initialize
+    @username = ENV['JIRA_USERNAME'] ||= DEFAULT_USERNAME
+    @password = ENV['JIRA_PASSWORD'] ||= File.read(format('%s/.wiki_pass', ENV['HOME'])).chomp
+    super
+  end
+
+  def get_url
+    'https://HOSTED_SITE.atlassian.net/wiki'
```

Comment:
What's `HOSTED_SITE`? Is the reader expected to replace these values themselves in the code? If so, it'd be helpful to make that clear above.

Reply:
I feel that this should be clear to anyone trying this out.  This would only apply to the hosted version of confluence, but in some cases people run it internally.  Ideally the "gold plated" solution would be to have this expressed as a global thingie.  I reserve the right to have some sucky code as I am ops and not amazeballs code god like y'all.

Reply:
It definitely would become clear that there are values that need to be replaced once you tried to use it -- I think it'd be helpful and avoid frustration to just mention it above the codeblock, like:

```
You should replace the following strings with your own values:
- HOSTED_SITE
- DEFAULT_USERNAME
...
```